src/upstream_temperature_feedback.cxx

```diff
@@ -0,0 +1,60 @@
+#include "../include/upstream_temperature_feedback.hxx"
+
+#include <bout/mesh.hxx>
+using bout::globals::mesh;
+
+void UpstreamTemperatureFeedback::transform(Options& state) {
+  Options& species = state["species"][name];
+
+  // Doesn't need all boundaries to be set
+  Field3D T = getNoBoundary<Field3D>(species["temperature"]);
+
+  auto time = get<BoutReal>(state["time"]);
+
+  for (RangeIterator r = mesh->iterateBndryLowerY(); !r.isDone(); r++) {
+    int jz = 0;
+    error = temperature_upstream - T(r.ind, mesh->ystart, jz);
+
+    // PI controller, using crude integral of the error
+    if (temperature_error_lasttime < 0.0) {
+      // First time this has run
+      temperature_error_lasttime = time;
+      temperature_error_last = error;
+    }
+
+    // Integrate using Trapezium rule
+    if (time > temperature_error_lasttime) { // Since time can decrease
+      temperature_error_integral += (time - temperature_error_lasttime) * 0.5 *
+        (error + temperature_error_last);
+    }
+
+    if ((temperature_error_integral < 0.0) && temperature_integral_positive) {
+      // Limit temperature_error_integral to be >= 0
+      temperature_error_integral = 0.0;
+    }
+
+    // Calculate source from combination of error and integral
+    integral_term = temperature_controller_i * temperature_error_integral;
+    proportional_term = temperature_controller_p * error;
+    source_multiplier = proportional_term + integral_term;
```

Comment:
Are we dropping the units of the temperature and time here? `integral_term` and `proptional_term` are meant to be dimensionless. Does that imply that `temperature_controller_i` has units of [time]^-1 * [temperature]^-1 and `temperature_controller_p` has units of [temperature]^-1?

Reply:
The P and I components are quantities calculated from dimensional data, but their own dimensionality doesn't matter because they're just control signals. I don't think there's any benefit in normalising them unless I'm missing something.

Reply:
On your other comments:
**Add in the "d" of a PID (example code below in Python)?**
I've raised this with Ben before as it seemed like a no brainer. There is a fundamental issue in calculating the D parameter, but I can't remember what it is. @bendudson?

After working with the density controller and tuning it on several occasions the P and I parameters seem more than sufficient. In fact you can get perfect performance with just the P parameter in some cases as well. I think this is because you are controlling a source which is in the same location as your control variable (i.e. the source and the control are both upstream) which makes the system response very fast and makes the control very easy.

**Is it possible to set Pe = Pd+, for Pd+ controlled via feedback or vice-versa? Probably want to have a Qi/Qe ratio ratio than a Ti/Te ratio**
Good question. I think this would need to be implemented in the code. One suggestion would be to add an option for a coupled species and a heat source ratio. You could add the controller for `d+` only, then set `e` as the coupled species and give it a ratio of 1. The controller then could pick this up from the input file and create a corresponding heat source for `e` calculated from the `d+` heat source times 1.